src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -9124,18 +9140,22 @@ size_t MergeTreeData::unloadPrimaryKeysOfOutdatedParts()
             /// Outdated part may be hold by SELECT query and still needs the index.
             /// This check requires lock of index_mutex but if outdated part is unique then there is no
             /// contention on it, so it's relatively cheap and it's ok to check under a global parts lock.
-            if (isSharedPtrUnique(part) && part->isIndexLoaded())
-                parts_to_unload_index.push_back(part);
+            if (isSharedPtrUnique(part) && (part->isIndexLoaded() || part->mayStoreDataInCaches()))
+                parts_to_clear.push_back(part);
         }
     }
 
-    for (const auto & part : parts_to_unload_index)
+    for (const auto & part : parts_to_clear)
     {
-        const_cast<IMergeTreeDataPart &>(*part).unloadIndex();
+        auto & part_mut = const_cast<IMergeTreeDataPart &>(*part);
+
+        part_mut.unloadIndex();
+        part_mut.clearCaches();
```

Comment:
Maybe a bit strange place to do it but it makes sense. I considered 2 alternatives:
- `removePartsFromWorkingSet` where we set `Outdated` state to data part. It's bad because it's called with locked parts lock and it's better not add more contention in that place by clearing some caches. Also some running selects may still benefit from cache.
- Desctructor of data part. It's bad because the outdated data will be too long in cache and it won't work well if there is a high rate of inserts with enabled prewarm.

So, it's here.